### ISSUE TITLE
Not working with 172.1.0.0/24 subnet

### ISSUE THREAD

**Issue by stigjoergensen** [klutchell/unbound-docker#48]
I just downloaded a clean version using the following compose file:

```
version: "3.5"

services:
  pihole:
    container_name: pihole
    image: pihole/pihole:latest
    ports:
      - "192.x.x.x:53:53/tcp"
      - "192.x.x.x:53:53/udp"
!     - "52:53/tcp"
!      - "52:53/udp"
!     - "67:67/udp"
      - "81:80/tcp"
    networks:
      default:
        ipv4_address: 172.1.0.3
    environment:
      TZ: "Europe/Copenhagen"
      PIHOLE_DNS: "172.1.0.2;172.1.0.2"
      WEBPASSWORD: 'xxxxxx'
    volumes:
      - '/docker/pihole/data/:/etc/pihole/'
      - '/docker/pihole/dnsmasq/:/etc/dnsmasq.d/'
    cap_add:
      - NET_ADMIN
    restart: unless-stopped

  unbound:
    container_name: unbound
    image: klutchell/unbound
    networks:
      default:
        ipv4_address: 172.1.0.2

networks:
  default:
    driver: bridge
    name: pihole
    ipam:
      config:
      - subnet: 172.1.0.0/24
!     - gateway: 172.1.0.1

```

(used ! instead of haskmark as that seems to do something in the editor)

using dig: dig @192.x.x.x -p 53 dr.dk 
i can see in the query log "OK, sent to unbound.pihole#53" that the request was send to unbound but the response never made it back

using docker exec -ti pihole /bin/bash im able from within the container to use 
dig @172.1.0.2 dr.dk
and get a response back from unbound.

in the pihole dns setting i have this setup
![image](https://user-images.githubusercontent.com/23189291/154846877-6ce0e70c-1268-4553-91d8-2a36b05256c2.png)

so it should be working correctly.

Any ideas?

**Comment by klutchell** [comment]
The formatting of your compose file is hard to read without three backticks (```) before and after.

Is anything printed in the unbound logs via `docker-compose logs unbound`?

**Comment by stigjoergensen** [comment]
Fixed the compose file.

ran the container again without -d to get a fresh and clean log

```
sudo docker-compose up
Creating network "pihole" with driver "bridge"
Creating pihole  ... done
Creating unbound ... done
Attaching to unbound, pihole
unbound    | [1645369523] unbound[1:0] notice: init module 0: validator
unbound    | [1645369523] unbound[1:0] notice: init module 1: iterator
unbound    | [1645369524] unbound[1:0] info: start of service (unbound 1.15.0).
unbound    | [1645369524] unbound[1:0] info: generate keytag query _ta-4a5c-4f66. NULL IN
pihole     | [s6-init] making user provided files available at /var/run/s6/etc...exited 0.
pihole     | [s6-init] ensuring user provided files have correct perms...exited 0.
pihole     | [fix-attrs.d] applying ownership & permissions fixes...
pihole     | [fix-attrs.d] 01-resolver-resolv: applying... 
pihole     | [fix-attrs.d] 01-resolver-resolv: exited 0.
pihole     | [fix-attrs.d] done.
pihole     | [cont-init.d] executing container initialization scripts...
pihole     | [cont-init.d] 05-changer-uid-gid.sh: executing... 
pihole     | [cont-init.d] 05-changer-uid-gid.sh: exited 0.
pihole     | [cont-init.d] 20-start.sh: executing... 
pihole     |  ::: Starting docker specific checks & setup for docker pihole/pihole
pihole     | 
pihole     |   [i] Installing configs from /etc/.pihole...
pihole     |   [i] Existing dnsmasq.conf found... it is not a Pi-hole file, leaving alone!
  [✓] Installed /etc/dnsmasq.d/01-pihole.conf
  [✓] Installed /etc/dnsmasq.d/06-rfc6761.conf
pihole     | Existing DNS servers detected in setupVars.conf. Leaving them alone
pihole     | ::: Pre existing WEBPASSWORD found
pihole     | DNSMasq binding to default interface: eth0
pihole     | Added ENV to php:
pihole     |                    "TZ" => "Europe/Copenhagen",
pihole     |                    "PIHOLE_DOCKER_TAG" => "2022.02.1",
pihole     |                    "PHP_ERROR_LOG" => "/var/log/lighttpd/error.log",
pihole     |                    "ServerIP" => "0.0.0.0",
pihole     |                    "CORS_HOSTS" => "",
pihole     |                    "VIRTUAL_HOST" => "0.0.0.0",
pihole     | Using IPv4 and IPv6
pihole     | ::: Preexisting ad list /etc/pihole/adlists.list detected ((exiting setup_blocklists early))
pihole     | https://raw.githubusercontent.com/StevenBlack/hosts/master/hosts
pihole     | ::: Testing lighttpd config: Syntax OK
pihole     | ::: All config checks passed, cleared for startup ...
pihole     | ::: Enabling Query Logging
pihole     |   [i] Enabling logging...
  [✓] Logging has been enabled!
pihole     |  ::: Docker start setup complete
pihole     |   Checking if custom gravity.db is set in /etc/pihole/pihole-FTL.conf
pihole     |   Pi-hole version is v5.9 (Latest: v5.9)
pihole     |   AdminLTE version is v5.11 (Latest: v5.11)
pihole     |   FTL version is v5.14 (Latest: v5.14)
pihole     |   Container tag is: 2022.02.1
pihole     | [cont-init.d] 20-start.sh: exited 0.
pihole     | [cont-init.d] done.
pihole     | [services.d] starting services
pihole     | [services.d] done.
pihole     | Starting pihole-FTL (no-daemon) as pihole
pihole     | Starting crond
pihole     | Starting lighttpd
```

**Comment by klutchell** [comment]
Thank you for cleaning up the compose file and logs, that's much easier to read!

> using docker exec -ti pihole /bin/bash im able from within the container to use
> dig @172.1.0.2 dr.dk
> and get a response back from unbound.

So it looks like unbound is working, but Pi-hole isn't responding to external requests.

> using dig: dig @192.x.x.x -p 53 dr.dk
> i can see in the query log "OK, sent to unbound.pihole#53" that the request was send to unbound but the response never made it back

Maybe we need to increase the debug logging of unbound. Can you add the following to your unbound service definition?
```
command: -vvv
```
This should increase the verbosity of the unbound logs and we can hopefully see if it receives the request from Pi-hole.

**Comment by stigjoergensen** [comment]
added the command so now the service looks like this:
```
  unbound:
    container_name: unbound
    image: klutchell/unbound
    command: -vvv
    networks:
      default:
        ipv4_address: 172.1.0.2
```

and we got a ton of log, but im not sure what to look for during the startup, but i do see this:

```
unbound    | [1645371742] unbound[1:0] debug: refused query from ip4 172.1.0.3 port 57768 (len 16)
unbound    | [1645371742] unbound[1:0] debug: refuse[54:0] 2313010000010000000000010234310138033136380331393207696E2D61646472046172706100000C000100002904B0000000000000
unbound    | [1645371747] unbound[1:0] debug: refused query from ip4 172.1.0.3 port 57768 (len 16)
unbound    | [1645371747] unbound[1:0] debug: refuse[54:0] 2313010000010000000000010234310138033136380331393207696E2D61646472046172706100000C000100002904D0000000000000
unbound    | [1645371752] unbound[1:0] debug: refused query from ip4 172.1.0.3 port 60162 (len 16)
unbound    | [1645371752] unbound[1:0] debug: refuse[51:0] F971010000010000000000010132013001310331373207696E2D61646472046172706100000C000100002904B0000000000000
unbound    | [1645371754] unbound[1:0] debug: refused query from ip4 172.1.0.3 port 38465 (len 16)
unbound    | [1645371754] unbound[1:0] debug: refuse[46:0] 38EB0120000100000000000102647202646B0000010001000029100000000000000C000A00083746338BE9EDBBB8
unbound    | [1645371757] unbound[1:0] debug: refused query from ip4 172.1.0.3 port 60162 (len 16)
unbound    | [1645371757] unbound[1:0] debug: refuse[51:0] F971010000010000000000010132013001310331373207696E2D61646472046172706100000C000100002904D0000000000000
unbound    | [1645371759] unbound[1:0] debug: refused query from ip4 172.1.0.3 port 38465 (len 16)
unbound    | [1645371759] unbound[1:0] debug: refuse[46:0] 38EB0120000100000000000102647202646B000001000100002904D000000000000C000A00083746338BE9EDBBB8
unbound    | [1645371764] unbound[1:0] debug: refused query from ip4 172.1.0.3 port 38465 (len 16)
unbound    | [1645371764] unbound[1:0] debug: refuse[46:0] 38EB0120000100000000000102647202646B000001000100002904D000000000000C000A00083746338BE9EDBBB8

```
have attached the complete log

[pihole.log](https://github.com/klutchell/unbound-docker/files/8104333/pihole.log)

**Comment by klutchell** [comment]
Hmm... I wonder if we need to allow that IP range to the allow list, or just change the IP range in the compose file.
https://github.com/klutchell/unbound-docker/blob/d7be321c9ccd55a726f248ad1285b391668444e0/rootfs_overlay/etc/unbound/unbound.conf#L74-L79

What if you change the pihole network to use addresses in the `10.x.x.x` or `172.16.x.x` range as a test?

**Comment by stigjoergensen** [comment]
it will not let me bring up in 10.x.x.x 

`ERROR: for pihole  Cannot start service pihole: Invalid address 10.1.1.3: It does not belong to any of this network's subnets`

even though the network is defined as:

```
networks:
  default:
    driver: bridge
    name: pihole
    ipam:
      config:
      - subnet: 10.1.1.0/24
```

Will try the 172.16 range now

**Comment by klutchell** [comment]
You might need to do a `docker-compose down` before bringing it up again so the new network is defined.

**Comment by stigjoergensen** [comment]
yep - that was the problem, even a --force-recreate didnt seem to remove the old network.

so changed the network to 172.16.1.0/24 and this seems to be working.. let me try on the 10/8 net again... just to 100% confirm thats the problem

**Comment by stigjoergensen** [comment]
yes, confirmed - changing the network to an allowed network defined in unbound seems to do the trick..   So what is the right way going forward? i can leave it as defined now... but i think your image should enter include a volume for unbound so people can change it, or you should include a settings and then add that to the allow list.

**Comment by klutchell** [comment]
> but i think your image should enter include a volume for unbound

Users are encouraged to mount their own configuration and they can use the included one as an example. There are instructions on the readme for this.

> you should include a settings and then add that to the allow list.

I have included a default settings file that allows [some private subnets](https://github.com/klutchell/unbound-docker/blob/main/rootfs_overlay/etc/unbound/unbound.conf#L74-L79) but `172.1.0.0/24` is not in the list of [reserved addresses](https://en.wikipedia.org/wiki/Reserved_IP_addresses). If users need more addresses they can open a PR or mount their own configuration file.

Glad we got it working, just make sure to use address ranges on the allowed subnet list or mount your own config!
